reactor-netty-core/src/main/java/reactor/netty/resources/PooledConnectionProvider.java

```diff
@@ -442,12 +444,21 @@ PoolBuilder<T, PoolConfig<T>> newPoolInternal(
 			PoolBuilder<T, PoolConfig<T>> poolBuilder =
 					PoolBuilder.from(allocator)
 					           .destroyHandler(destroyHandler)
-					           .evictionPredicate(evictionPredicate
-					                   .or((poolable, meta) -> (maxIdleTime != -1 && meta.idleTime() >= maxIdleTime)
-					                           || (maxLifeTime != -1 && meta.lifeTime() >= maxLifeTime)))
 					           .maxPendingAcquire(pendingAcquireMaxCount)
 					           .evictInBackground(evictionInterval);
 
+			BiPredicate<T, PooledRefMetadata> poolEvictionPredicate = evictionPredicate
+					.or((poolable, meta) -> (maxIdleTime != -1 && meta.idleTime() >= maxIdleTime)
+							|| (maxLifeTime != -1 && meta.lifeTime() >= maxLifeTime));
+
+			if (this.evictionPredicate != null) {
+				poolBuilder = poolBuilder.evictionPredicate(poolEvictionPredicate
+						.or((poolable, meta) -> this.evictionPredicate.test(poolable, new PooledConnectionMetadata(meta))));
```

Comment:
@violetagg let me know if you're happy with the extra allocation

Reply:
Good point - agree we should fully override the default if user supplied.

Reply:
Do you agree maxLifeTime and maxIdleTime should be applied in addition to either the default or user-supplied eviction predicate if they are specified?

Reply:
hm yeah ... I don't know ... `AllocationStrategy` for example disables the max connection setting

Also are we targeting HTTP/2 because the pool implementation there is more complicated from connections/streams handling point of view and the eviction is not trivial.

reactor.netty.http.client.Http2Pool#evictInBackground
reactor.netty.http.client.Http2Pool#destroyPoolable
reactor.netty.http.client.Http2Pool#findConnection

Reply:
It does make sense in my mind to also target HTTP/2. What do you think?

Reply:
yes

You may need to rebase this work on top of the latest changes (in case you see tests failing)

Reply:
Hooked into HTTP2 pool now.